File: objectiveai-rs/src/agent/completions/request/response_format.rs
//! Response format configuration for agent completions.

use std::hash::{Hash, Hasher};
use indexmap::IndexMap;
use serde::{Deserialize, Serialize};
use schemars::JsonSchema;

/// The format of the model's response.
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
#[serde(tag = "type", rename_all = "snake_case")]
#[schemars(rename = "agent.completions.request.ResponseFormat")]
pub enum ResponseFormat {
    /// Plain text response (default).
    Text,
    /// Response must be valid JSON.
    JsonObject,
    /// Response must conform to a JSON schema.
    JsonSchema {
        /// The JSON Schema definition.
        schema: IndexMap<String, serde_json::Value>,
    },
    /// Response must conform to a grammar.
    Grammar { grammar: String },
    /// Response must be valid Python code.
    Python,
    /// The final assistant message will contain this tool call
    ToolCall {
        /// The name of the tool.
        name: String,
        /// A description of the tool.
        description: String,
        /// The JSON Schema definition.
        schema: IndexMap<String, serde_json::Value>,
        /// Whether the tool MUST be called.
        #[serde(skip_serializing_if = "Option::is_none")]
        required: Option<bool>,
    },
}

impl ResponseFormat {
    pub fn tool_key(&self) -> String {
        let json = serde_json::to_string(self).unwrap_or_default();
        let mut h = std::collections::hash_map::DefaultHasher::new();
        json.hash(&mut h);
        format!("{:x}", h.finish())
    }
}

/// A map from agent ID to response format, allowing per-agent configuration.
pub type PerAgentResponseFormat = IndexMap<String, ResponseFormat>;

/// Either a single response format or a per-agent map.
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
#[serde(untagged)]
#[schemars(rename = "agent.completions.request.ResponseFormatParam")]
pub enum ResponseFormatParam {
    /// A single response format applied to all agents.
    Single(ResponseFormat),
    /// Per-agent response formats, keyed by agent ID.
    PerAgent(PerAgentResponseFormat),
}

File: objectiveai-js/src/agent/completions/request/responseFormat.ts
import { z } from "zod";

export const AgentCompletionsRequestResponseFormatSchema = z.union([z.object({
  type: z.literal("text"),
}).describe("Plain text response (default)."), z.object({
  type: z.literal("json_object"),
}).describe("Response must be valid JSON."), z.object({
  schema: z.record(z.string(), z.unknown()).describe("The JSON Schema definition."),
  type: z.literal("json_schema"),
}).describe("Response must conform to a JSON schema."), z.object({
  grammar: z.string(),
  type: z.literal("grammar"),
}).describe("Response must conform to a grammar."), z.object({
  type: z.literal("python"),
}).describe("Response must be valid Python code."), z.object({
  name: z.string().describe("The name of the tool."),
  description: z.string().describe("A description of the tool."),
  schema: z.record(z.string(), z.unknown()).describe("The JSON Schema definition."),
  required: z.boolean().nullable().describe("Whether the tool MUST be called.").optional(),
  type: z.literal("tool_call"),
}).describe("The final assistant message will contain this tool call")]).describe("The format of the model's response.").meta({ title: "agent.completions.request.ResponseFormat" });
export type AgentCompletionsRequestResponseFormat = z.infer<typeof AgentCompletionsRequestResponseFormatSchema>;
